doc/man/gnunet-update.1
.TH GNUNET-UPDATE "1" "23 Dec 2006" "GNUnet"
.SH NAME
gnunet-update \- a tool to update GNUnet
.SH SYNOPSIS
.B gnunet\-update
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
gnunet-update updates gnunets databases after software updates
that change the database format.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (default: /etc/gnunet.conf)
.TP
\fB\-g \fISECTION:ENTRY\fR, \fB\-\-get=SECTION:ENTRY\fR
print the value from the configuration in section SECTION under entry ENTRY.  Use :ENTRY to access the global section.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).
.TP
\fB\-U\fR, \fB\-\-client\fR
run in client mode (required to get values from the client configuration).  This option must be used only together with the \-g option.
.TP
\fB\-u \fIUSER\fR, \fB\-\-user=USER\fR
run as user USER (and if available as group USER). Note that to use this option, you will probably have to start gnunet-update as
root. It is typically better to directly start gnunet-update as that user instead.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR \fB\-\-verbose\fR
Be verbose.
.SH NOTES
gnunet-update was introduced in version 0.6.3.  Updates from
much earlier versions may not be handled properly by the tool.
Always consult the UPDATING file in the GNUnet distribution
for the necessary update procedure.
.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunetd\fP(1), \fBgnunet\-peer\-info\fP(1), \fBgnunet\-transport\-check\fP(1), \fBgnunet.conf\fP(5)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/server/gnunet-update.c
/*
     This file is part of GNUnet.
     (C) 2004, 2006, 2008 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file server/gnunet-update.c
 * @brief tool to process changes due to version updates
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_util.h"
#include "gnunet_directories.h"
#include "gnunet_core.h"
#include "core.h"
#include "startup.h"
#include "version.h"

/**
 * We may want to change this at some point into
 * something like libgnunet_update if we want to
 * separate the update code from the codebase
 * used in normal operation -- but currently I
 * see no need / use for that.
 */
#define DSO_PREFIX "libgnunet"

static struct GNUNET_GC_Configuration *cfg;

static struct GNUNET_GE_Context *ectx;

static char **processed;

static unsigned int processedCount;

static GNUNET_UpdateAPI uapi;

static char *cfgFilename = GNUNET_DEFAULT_DAEMON_CONFIG_FILE;


/**
 * Allow the module named "pos" to update.
 * @return GNUNET_OK on success, GNUNET_SYSERR on error
 */
static int
updateModule (const char *rpos)
{
  GNUNET_UpdatePluginMainMethod mptr;
  struct GNUNET_PluginHandle *library;
  char *name;
  int i;
  char *pos;

  for (i = 0; i < processedCount; i++)
    if (0 == strcmp (rpos, processed[i]))
      {
        return GNUNET_OK;       /* already done */
      }
  GNUNET_array_grow (processed, processedCount, processedCount + 1);
  processed[processedCount - 1] = GNUNET_strdup (rpos);

  pos = NULL;
  if (-1 == GNUNET_GC_get_configuration_value_string (cfg,
                                                      "MODULES", rpos, rpos,
                                                      &pos))
    return GNUNET_SYSERR;
  GNUNET_GE_ASSERT (ectx, pos != NULL);

  name = GNUNET_malloc (strlen (pos) + strlen ("module_") + 1);
  strcpy (name, "module_");
  strcat (name, pos);
  GNUNET_free (pos);
  library = GNUNET_plugin_load (ectx, DSO_PREFIX, name);
  if (library == NULL)
    {
      GNUNET_free (name);
      return GNUNET_SYSERR;
    }
  mptr = GNUNET_plugin_resolve_function (library, "update_", GNUNET_NO);
  if (mptr == NULL)
    {
      GNUNET_plugin_unload (library);
      GNUNET_free (name);
      return GNUNET_OK;         /* module needs no updates! */
    }
  mptr (&uapi);
  GNUNET_plugin_unload (library);
  GNUNET_free (name);
  return GNUNET_OK;
}

/**
 * Call the update module for each of the applications
 * in the current configuration.
 */
static void
updateApplicationModules ()
{
  char *dso;
  char *next;
  char *pos;

  dso = NULL;
  if (-1 == GNUNET_GC_get_configuration_value_string (cfg,
                                                      "GNUNETD",
                                                      "APPLICATIONS",
                                                      "advertising fs getoption stats traffic",
                                                      &dso))
    return;
  GNUNET_GE_ASSERT (ectx, dso != NULL);
  next = dso;
  do
    {
      pos = next;
      while ((*next != '\0') && (*next != ' '))
        next++;
      if (*next == '\0')
        {
          next = NULL;          /* terminate! */
        }
      else
        {
          *next = '\0';         /* add 0-termination for pos */
          next++;
        }
      if (strlen (pos) > 0)
        {
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_INFO | GNUNET_GE_USER | GNUNET_GE_BULK,
                         _("Updating data for module `%s'\n"), pos);
          if (GNUNET_OK != updateModule (pos))
            GNUNET_GE_LOG (ectx,
                           GNUNET_GE_ERROR | GNUNET_GE_DEVELOPER |
                           GNUNET_GE_BULK | GNUNET_GE_USER,
                           _("Failed to update data for module `%s'\n"), pos);
        }
    }
  while (next != NULL);
  GNUNET_free (dso);
}

static void
doGet (char *get)
{
  char *sec;
  char *ent;
  char *val;

  sec = get;
  ent = get;
  while (((*ent) != ':') && ((*ent) != '\0'))
    ent++;
  if (*ent == ':')
    {
      *ent = '\0';
      ent++;
    }
  if (GNUNET_YES == GNUNET_GC_have_configuration_value (cfg, sec, ent))
    {
      GNUNET_GC_get_configuration_value_string (cfg, sec, ent, NULL, &val);
      printf ("%s\n", val);
      GNUNET_free (val);
    }
}

static void
work ()
{
  int i;
  struct GNUNET_CronManager *cron;
  char *topo;

  uapi.service_update = &updateModule;
  uapi.service_request = &GNUNET_CORE_request_service;
  uapi.service_release = &GNUNET_CORE_release_service;
  uapi.ectx = ectx;
  uapi.cfg = cfg;
  /* have no client disturb us! */
  GNUNET_GC_set_configuration_value_string (cfg,
                                            NULL,
                                            "TCPSERVER", "DISABLE", "YES");
  GNUNET_GC_get_configuration_value_string (cfg,
                                            "MODULES",
                                            "topology",
                                            "topology_default", &topo);
  /* Code specific for update from 0.7.2c to 0.7.3 */
  if (0 == strcmp (topo, "topology_f2f"))
    {
      GNUNET_GC_set_configuration_value_string (cfg,
                                                ectx,
                                                "MODULES",
                                                "topology",
                                                "topology_default");
      GNUNET_GC_set_configuration_value_string (cfg,
                                                ectx,
                                                "F2F", "FRIENDS-ONLY", "YES");
      if (GNUNET_OK == GNUNET_GC_write_configuration (cfg, cfgFilename))
        fprintf (stdout, "Updated F2F configuration options successfully.\n");
      else
        fprintf (stdout,
                 "Failed to write configuration with updated F2F configuration.\n");
    }
  GNUNET_free (topo);
  cron = GNUNET_cron_create (ectx);
  if (GNUNET_CORE_init (ectx, cfg, cron, NULL) != GNUNET_OK)
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_FATAL | GNUNET_GE_USER | GNUNET_GE_IMMEDIATE,
                     _("Core initialization failed.\n"));

      return;
    }



  /* enforce filesystem limits */
  GNUNET_CORE_startup_cap_fs_quota_size (ectx, cfg);

  /* force update of common modules (used by core) */
  updateModule ("transport");
  updateModule ("identity");
  updateModule ("session");
  updateModule ("fragmentation");
  updateModule ("topology");
  /* then update active application modules */
  updateApplicationModules ();
  /* store information about update */
  GNUNET_CORE_version_mark_as_up_to_date (ectx, cfg);

  for (i = 0; i < processedCount; i++)
    GNUNET_free (processed[i]);
  GNUNET_array_grow (processed, processedCount, 0);
  GNUNET_CORE_done ();
  GNUNET_cron_destroy (cron);
}

static int
set_client_config (GNUNET_CommandLineProcessorContext * ctx,
                   void *scls, const char *option, const char *value)
{
  cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;
  return GNUNET_OK;
}


/**
 * All gnunet-update command line options
 */
static struct GNUNET_CommandLineOption gnunetupdateOptions[] = {
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'g', "get", "SECTION:ENTRY",
   gettext_noop ("print a value from the configuration file to stdout"),
   1, &GNUNET_getopt_configure_set_option, "GNUNET-UPDATE:GET"},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Updates GNUnet datastructures after version change.")),       /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'u', "user", "LOGIN",
   gettext_noop ("run as user LOGIN"),
   1, &GNUNET_getopt_configure_set_option, "GNUNETD:USER"},
  {'U', "client", NULL,
   gettext_noop
   ("run in client mode (for getting client configuration values)"),
   0, &set_client_config, NULL},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};


int
main (int argc, char *const *argv)
{
  char *get;
  int ret;

  ret = GNUNET_init (argc,
                     argv,
                     "gnunet-update",
                     &cfgFilename, gnunetupdateOptions, &ectx, &cfg);
  if ((ret == -1)
      || (GNUNET_OK != GNUNET_CORE_startup_change_user (ectx, cfg)))
    {
      GNUNET_fini (ectx, cfg);
      return -1;
    }
  get = NULL;
  GNUNET_GC_get_configuration_value_string (cfg, "GNUNET-UPDATE", "GET", "",
                                            &get);
  if (strlen (get) > 0)
    doGet (get);
  else
    work ();
  GNUNET_free (get);
  GNUNET_fini (ectx, cfg);
  return 0;
}

/* end of gnunet-update */
